app/Platform/Controllers/CompareUnlocksController.php

```diff
@@ -0,0 +1,120 @@
+                $achievements[$unlock->achievement_id]['userHardcore'] = false;
+            }
+        }
+
+        $otherUserUnlocks = $user->playerAchievements()
+            ->join('Achievements', 'Achievements.ID', '=', 'player_achievements.achievement_id')
+            ->where('Achievements.GameID', $game->ID)
+            ->select(['achievement_id', 'unlocked_at', 'unlocked_hardcore_at']);
+        foreach ($otherUserUnlocks->get() as $unlock) {
+            if (!array_key_exists($unlock->achievement_id, $achievements)) {
+                continue;
+            }
+
+            if ($unlock->unlocked_hardcore_at) {
+                $achievements[$unlock->achievement_id]['otherUserTimestampRaw'] = $unlock->unlocked_hardcore_at;
+                $achievements[$unlock->achievement_id]['otherUserTimestamp'] = Carbon::parse($unlock->unlocked_hardcore_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['otherUserHardcore'] = true;
+            } elseif ($unlock->unlocked_at) {
+                $achievements[$unlock->achievement_id]['otherUserTimestampRaw'] = $unlock->unlocked_at;
+                $achievements[$unlock->achievement_id]['otherUserTimestamp'] = Carbon::parse($unlock->unlocked_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['otherUserHardcore'] = false;
+            }
+        }
+
+        $this->sortList($achievements, $sortOrder);
+
+        return view('pages.user.[user].game.[game].compare', [
+            'user' => $activeUser,
+            'otherUser' => $user,
+            'game' => $game,
+            'achievements' => $achievements,
+            'sortOrder' => $sortOrder,
+        ]);
+    }
+
+    private function sortList(array &$achievements, string $sortOrder): void
+    {
+        $sortFunction = match ($sortOrder) {
+            default => function ($a, $b) {
+                return $this->sortByUnlockTimestamps($a, $b, 'userTimestampRaw');
+            },
+            'otherUnlocks' => function ($a, $b) {
+                return $this->sortByUnlockTimestamps($a, $b, 'otherUserTimestampRaw');
+            },
+            'display' => function ($a, $b) {
+                return $a['DisplayOrder'] <=> $b['DisplayOrder'];
+            },
+            'title' => function ($a, $b) {
+                return $a['Title'] <=> $b['Title'];
+            },
+        };
+
+        usort($achievements, $sortFunction);
+    }
+
+    private function sortByUnlockTimestamps(array $a, array $b, string $field): int
+    {
+        $aTimestamp = $a[$field] ?? null;
+        $bTimestamp = $b[$field] ?? null;
+        if ($aTimestamp != $bTimestamp) {
+            if ($aTimestamp === null) {
+                return 1;
+            } elseif ($bTimestamp == null) {
+                return -1;
+            } else {
+                return $aTimestamp <=> $bTimestamp;
+            }
+        }
+
+        return $a['DisplayOrder'] <=> $b['DisplayOrder'];
+    }
```

Comment:
```php
private function sortByUnlockTimestamps(array $a, array $b, string $field): int
{
    // Force null values to be sorted last.
    $nullTimestampValue = '9999-01-01 12:00:00';

    $aTimestamp = $a[$field] ?? $nullTimestampValue;
    $bTimestamp = $b[$field] ?? $nullTimestampValue;

    if ($aTimestamp !== $bTimestamp) {
        return $aTimestamp <=> $bTimestamp;
    }

    // If timestamps are equal, fall back to the DisplayOrder value.
    return $a['DisplayOrder'] <=> $b['DisplayOrder'];
}
```

Functionally equivalent, and whittles the implementation down from 4 possible returns to 2.